man/systemd-soft-reboot.service.xml

```diff
@@ -0,0 +1,155 @@
+      the same unit. For this to work, units must declare <varname>DefaultDependencies=no</varname> (and
+      avoid a manual <varname>Conflicts=shutdown.target</varname> or similar) to ensure they are not
+      terminated as usual during the system shutdown operation. Alternatively, use
+      <varname>FileDescriptorStorePreserve=</varname> to allow the file descriptor store to remain pinned
+      even when the unit is down. See
+      <citerefentry><refentrytitle>systemd.service</refentrytitle><manvolnum>5</manvolnum></citerefentry> for
+      details about the file descriptor store.</para></listitem>
+
+      <listitem><para>Similar to this, file descriptors associated with <filename>.socket</filename> units
+      remain open (and connectible) if the units are not stopped during the transition. (Achieved by
+      <varname>DefaultDependencies=no</varname>.)</para></listitem>
+
+      <listitem><para>The <filename>/run/</filename> file system remains mounted and populated and may be
+      used to pass state information between such userspace reboot cycles.</para></listitem>
+
+      <listitem><para>Service processes may continue to run over the transition, if they are placed in
+      services that remain active until the very end of shutdown (which again is achieved via
+      <varname>DefaultDependencies=no</varname>). They must also be set up to avoid being killed by the
+      aforementioned <constant>SIGTERM</constant> spree (as per <ulink
+      url="https://systemd.io/ROOT_STORAGE_DAEMONS">systemd and Storage Daemons for the Root File
+      System</ulink>).</para></listitem>
+
+      <listitem><para>File system mounts may remain mounted during the transition, and complex storage
+      attached, if configured to remain until the very end of the shutdown process. (Also achieved via
+      <varname>DefaultDependencies=no</varname>, and by avoiding
+      <varname>Conflicts=umount.target</varname>)</para></listitem>
+    </itemizedlist>
+
+    <para>Even though passing resources from one soft reboot cycle to the next is possible this way we
+    strongly suggest using this functionality sparingly, as it creates a more fragile system as resources
+    from different versions of the OS and applications might be mixed with unforeseen consequences. In
+    particular it's recommended to avoid allowing processes to survive the soft reboot operation, as this
+    means code updates will necessarily be incomplete, and processes typically pin various other resources
+    (such as the file system they are backed by), thus increasing memory usage (as two versions of the
+    OS/application/file system might be kept in memory). Leaving processes running during a soft-reboot
+    operation requires disconnecting the service comprehensively from the rest of the OS, i.e. minimizing IPC
+    and reducing sharing of resources with the rest of the OS.</para>
```

Comment:
maybe plug in portable services and nspawn as solutions that work well with this, and link their docs?

Reply:
nspawn isn't that great for this, because it runs a supervisor process for which we (currently) have no good concept of updating. portable services otoh do not have that (their supervisor is pid1 after all, and that we do know how to replace). hence added a mention of those, but left nay mention of nspawn out for now.

Reply:
Yeah, we think about how to do something about nspawn to make it work nicely in this use case, separately of course - maybe implement daemon-reexec for it too?

Reply:
yeah, reexec would probably make sense